src/main/java/li/cil/tis3d/client/gui/CodeBookGui.java

```diff
@@ -631,41 +634,42 @@ private void drawProgram(final int mouseX, final int mouseY) {
                 rawEndX = columnToX(localLineNumber, exception.getEnd());
             }
             final int startY = guiY + CODE_POS_Y + localLineNumber * getTextRenderer().fontHeight - 1;
-            final int endX = Math.max(rawEndX, startX + getTextRenderer().getStringWidth(" "));
+            final int endX = Math.max(rawEndX, startX + getTextRenderer().getWidth(" "));
 
-            fill(startX - 1, startY + getTextRenderer().fontHeight - 1, endX, startY + getTextRenderer().fontHeight, 0xFFFF3333);
+            fill(matrices, startX - 1, startY + getTextRenderer().fontHeight - 1, endX, startY + getTextRenderer().fontHeight, 0xFFFF3333);
 
             // Draw selection position in text.
-            drawTextCursor();
+            drawTextCursor(matrices);
 
             // Part two of error handling, draw tooltip, *on top* of blinking cursor.
             if (mouseX >= startX && mouseX <= endX && mouseY >= startY && mouseY <= startY + getTextRenderer().fontHeight) {
-                final List<String> tooltip = new ArrayList<>();
+                final List<StringRenderable> tooltip = new ArrayList<>();
                 if (isErrorOnPreviousPage) {
-                    tooltip.add(I18n.translate(Constants.MESSAGE_ERROR_ON_PREVIOUS_PAGE));
+                    tooltip.add(FontRendererUtils.translate(Constants.MESSAGE_ERROR_ON_PREVIOUS_PAGE));
                 } else if (isErrorOnNextPage) {
-                    tooltip.add(I18n.translate(Constants.MESSAGE_ERROR_ON_NEXT_PAGE));
+                    tooltip.add(FontRendererUtils.translate(Constants.MESSAGE_ERROR_ON_NEXT_PAGE));
                 }
-                tooltip.addAll(Arrays.asList(Constants.PATTERN_LINES.split(I18n.translate(exception.getMessage())))
-                );
-                renderTooltip(tooltip, mouseX, mouseY);
+                final String translatedException = I18n.translate(exception.getMessage());
+                final List<String> lines = Arrays.asList(Constants.PATTERN_LINES.split(translatedException));
+                FontRendererUtils.addStringListToTooltip(lines, tooltip);
```

Comment:
Since it's the only usage (and foreseeably remaining so?) I'd probably inline `addStringListToTooltip` here? Your choice.

Alternative way of going about this might be a custom `Text` implementation that wraps a `new TranslatedText(exception.getMessage())` and does the splitting in its `visit`?

Reply:
yep, inlined it locally.